packages/dbml-core/src/parse/dbml/src/compiler.ts

```diff
@@ -1,52 +1,535 @@
+          return new Some(id - 1);
+        },
+      ),
+      // Return the index in the flatStream of the first token after/containing the offset
+      afterOrContain: this.createQuery(
+        Query.Token_NonTrivial_AfterOrContain,
+        (offset: number): Option<number> => {
+          if (this.token.flatStream().length === 0) {
+            return new None();
+          }
 
-        return analyzer.analyze();
-      });
-  }
+          const id = this.token.nonTrivial.beforeOrContain(offset).unwrap_or(-1);
+          if (id === -1) {
+            return new Some(0);
+          }
+          if (isOffsetWithinSpan(offset, this.token.flatStream()[id])) {
+            return new Some(id);
+          }
+
+          return id === this.token.flatStream().length ? new None() : new Some(id + 1);
+        },
+      ),
+      // Return the index in the flatStream of the last token before the offset
+      lastBefore: this.createQuery(
+        Query.Token_NonTrivial_BeforeOrContain,
+        (offset: number): Option<number> => {
+          const id = this.token.nonTrivial.afterOrContain(offset).unwrap_or(-1);
+
+          if (id === -1) {
+            const len = this.token.flatStream().length;
+
+            return len === 0 ? new None() : new Some(len - 1);
+          }
+
+          return id === 0 ? new None() : new Some(id - 1);
+        },
+      ),
+      // Return the index in the flatStream of the first token after the offset
+      firstAfter: this.createQuery(
+        Query.Token_NonTrivial_AfterOrContain,
+        (offset: number): Option<number> => {
+          const id = this.token.nonTrivial.beforeOrContain(offset).unwrap_or(-1);
+          const len = this.token.flatStream().length;
+
+          if (id === -1) {
+            return len === 0 ? new None() : new Some(0);
+          }
+
+          return id === len - 1 ? new None() : new Some(id);
+        },
+      ),
+      beforeOrContainOnSameLine: this.createQuery(
+        Query.Token_NonTrivial_FirstOfLine,
+        (offset: number): Option<number> => {
+          const id = this.token.nonTrivial.beforeOrContain(offset).unwrap_or(-1);
+          if (id === -1) {
+            return new None();
+          }
+          const token = this.token.flatStream()[id];
+          if (!isOffsetWithinSpan(offset, token)) {
+            const newline = token.trailingTrivia.find(
+              ({ kind }) => kind === SyntaxTokenKind.NEWLINE,
+            );
+            if (!newline || newline.start > offset) {
+              return new Some(id);
+            }
+
+            return new None();
+          }
+
+          return new Some(id);
+        },
+      ),
+    },
+  };
+
+  // A namespace for parsing-related utility
+  readonly parse = {
+    _: this.createQuery(
+      Query._Interpret,
+      (): Report<
+        Readonly<{ ast: ProgramNode; tokens: SyntaxToken[]; rawDb?: Database }>,
+        CompileError
+      > => {
+        const parseRes = new Lexer(this.source)
```

Comment:
One optimization is to divide this query into smaller sub-queries such as parse and analyze as in some cases of editor support (i.e. go to definition, go to references), running interpreter is unnecessary. This would cater for the case of in the future, we want to run language server in multiple workers and in each worker it only compute for things it needs.

Reply:
I did want to do that to, however, currently the interpreter would append more errors to `errors`, which means if I split it to another query, the cached `errors` would be mutated if I later call the interpreter after the parser. 🤔 I haven't come up with an appropriate solution. maybe we can split parsing errors and interpreter errors?